tests/components/icloud/test_config_flow.py

```diff
@@ -0,0 +1,308 @@
+"""Tests for the iCloud config flow."""
+import logging
+from unittest.mock import patch, MagicMock
+import pytest
+
+from pyicloud import PyiCloudService
+from pyicloud.exceptions import PyiCloudFailedLoginException
+
+from homeassistant import data_entry_flow
+from homeassistant.components.icloud import config_flow
+from homeassistant.components.icloud.config_flow import (
+    CONF_TRUSTED_DEVICE,
+    CONF_VERIFICATION_CODE,
+)
+from homeassistant.components.icloud.const import (
+    DOMAIN,
+    CONF_ACCOUNT_NAME,
+    CONF_GPS_ACCURACY_THRESHOLD,
+    CONF_MAX_INTERVAL,
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DEFAULT_MAX_INTERVAL,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+
+from tests.common import MockConfigEntry
+
+_LOGGER = logging.getLogger(__name__)
+
+USERNAME = "username@me.com"
+PASSWORD = "password"
+ACCOUNT_NAME = "Account name 1 2 3"
+ACCOUNT_NAME_FROM_USERNAME = None
+MAX_INTERVAL = 15
+GPS_ACCURACY_THRESHOLD = 250
+
+TRUSTED_DEVICES = [
+    {"deviceType": "SMS", "areaCode": "", "phoneNumber": "*******58", "deviceId": "1"}
+]
+
+
+def mock_icloud_service_with_cookie():
+    """Return a mocked iCloud service."""
+    service = MagicMock(
+        spec=PyiCloudService, requires_2fa=False, trusted_devices=TRUSTED_DEVICES
+    )
+    service.return_value.authenticate.return_value = None
+    return service
```

Comment:
Responding to https://github.com/home-assistant/home-assistant/pull/26590#discussion_r346722928

Should it be something like this @MartinHjelmare ?

Compared to your `mysensor.MQTTGateway` class : https://github.com/theolind/pymysensors/blob/master/mysensors/gateway_mqtt.py#L12-L25

the PyiCloudService doesn't have those properties in parameters : https://github.com/picklepete/pyicloud/blob/9bb6d750662ce24c8febc94807ddbdcdf3cadaa2/pyicloud/base.py#L125-L139

but in `@property` https://github.com/picklepete/pyicloud/blob/9bb6d750662ce24c8febc94807ddbdcdf3cadaa2/pyicloud/base.py#L239-L247

How can I do ?
Thanks

Reply:
You can set attributes on the mock instance to whatever you need:
https://github.com/MartinHjelmare/home-assistant/blob/a0397f46f88f0cec99a2f1f9476d5ef30aee05b7/tests/components/mysensors/common.py#L41

If you want to mimic the behavior of a property more tightly there's the `PropertyMock`:
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.PropertyMock

I would try to just set the attribute on the mock directly first.

The mock accepts arbitrary keyword arguments, when creating the mock, that when passed will be set as attributes on the mock:
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.configure_mock

Reply:
Okey, I get how to mock a property but my service still retrieve a MagicMock ...

<img width="1680" alt="Capture d’écran 2019-11-17 à 19 23 53" src="https://user-images.githubusercontent.com/14821482/69011961-99d3c080-0970-11ea-86f3-6fcf47fd99ba.png">